### PR TITLE
Update for GCC 12.2 on Ubuntu 22.04

### PR DESCRIPTION
**Pull request by rleh**

### CODE REVIEW

**Comment by rleh** [comment]
Seems to work 😱

File: build.sh [hunk 1]
```diff
@@ -9,23 +9,23 @@ mkdir ${BUILD}
...
-VERSION_LIBC="2.0.0"
+#VERSION_BINUTILS="2.38"
+VERSION_GCC="12.1.0"
+VERSION_LIBC="2.1.0"
 
 # Get sources
-wget -q   "https://raw.githubusercontent.com/archlinux/svntogit-community/c3efadcb76f4d8b1a3784015e7c472f59dbfa7de/avr-binutils/repos/community-x86_64/avr-size.patch" &
-wget -q   "https://raw.githubusercontent.com/osx-cross/homebrew-avr/d2e2566b06b90355952ed996707a0a1a24673cd3/Patch/avr-libc-add-mcu-atmega168pb.patch" &
+##wget -q   "https://raw.githubusercontent.com/archlinux/svntogit-community/c3efadcb76f4d8b1a3784015e7c472f59dbfa7de/avr-binutils/repos/community-x86_64/avr-size.patch" &
+##wget -q   "https://raw.githubusercontent.com/osx-cross/homebrew-avr/d2e2566b06b90355952ed996707a0a1a24673cd3/Patch/avr-libc-add-mcu-atmega168pb.patch" &
```
Comment (chris-durand) [inline build.sh]: The patches are not required anymore?

Comment (rleh) [reply build.sh]: Seems to work without the patches. I'll investigate further next week...

Comment (chris-durand) [reply build.sh]: The [archlinux build](https://github.com/archlinux/svntogit-community/tree/9605e9f7cf9d2aad9805659dff9866f0f72f215b/trunk) of avr-binutils 2.39 still uses the size patch. It is required for [some extra functionality](https://bugs.archlinux.org/task/32504) in `avr-size`. I will update the binutils version to 2.39 and add the patch back in.

Comment (chris-durand) [reply build.sh]: I have also added the atmega168pb patch again. The patches were applying fine, but there was an error in the `patch` command.

Next step is to test the compiled avr-gcc 12.2

**Review by salkinium — APPROVED** [review]
Looks perfectly reasonable to me, I'll port this verbatim to homebrew and test it on my mac.

**Comment by chris-durand** [comment]
The pipeline never ran for the last commit because of a permissions issue related to an external action. It is enabled in the repository settings. I guess it needs a force-push to try again.

**Review by chris-durand — APPROVED** [review]
The compiler seems to work. It emits tons of warnings for register accesses, though:

```
modm/src/modm/platform/uart/uart0_tx.cpp: In static member function 'static std::size_t modm::platform::Uart0::discardTransmitBuffer()':
modm/src/modm/platform/uart/uart0_tx.cpp:125:24: warning: array subscript 0 is outside array bounds of 'volatile uint8_t [0]' [-Warray-bounds]
  125 |                 UCSR0B &= ~(1 << UDRIE0);
      |                        ^
```
That is most likely a problem with the device headers included with the compiler. I have no AVR hardware to test if the generated binaries really work.

Merge?

**Comment by salkinium** [comment]
Lemme see if I can test it somehow, I couldn't compile this yet on macOS due to toolchain issues on ARM64.

**Comment by salkinium** [comment]
I still did not find a way to fix compilation on macOS, but that shouldn't stop this for Linux.

**Comment by salkinium** [comment]
I was able to build avr-gcc using x86 mode on ARM64, and the compiler seems to work fine for most simple apps, however, we may need to update our avr libstdc++ for GCC12, since there are several bugs related to libc.

For example, the `examples/avr/mega_pro` calls `sinf`, which is implemented in our libstdc++ by forwarding to libc `::sin`, which however, seems to never be called in the firmware, leading to an infinite loop. Here is the disassembly where the the missing call to `::sin` should be at address `310` but isn't (why?).

```
000002fc <sinf>:
sinf():
./modm/ext/gcc/libstdc++/src/math.cc:24
  {
    return ::cos(x);
  }

  float sinf(float x)
  {
 2fc:	cf 93       	push	r28
 2fe:	df 93       	push	r29
 300:	00 d0       	rcall	.+0      	; 0x302 <L0^A>

00000302 <L0^A>:
 302:	0f 92       	push	r0
 304:	cd b7       	in	r28, 0x3d	; 61
 306:	de b7       	in	r29, 0x3e	; 62

00000308 <.Loc.8>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 308:	69 83       	std	Y+1, r22	; 0x01
 30a:	7a 83       	std	Y+2, r23	; 0x02
 30c:	8b 83       	std	Y+3, r24	; 0x03
 30e:	9c 83       	std	Y+4, r25	; 0x04
 310:	69 81       	ldd	r22, Y+1	; 0x01
 312:	7a 81       	ldd	r23, Y+2	; 0x02
 314:	8b 81       	ldd	r24, Y+3	; 0x03
 316:	9c 81       	ldd	r25, Y+4	; 0x04

00000318 <.LVL4>:
./modm/ext/gcc/libstdc++/src/math.cc:26
  }
 318:	0f 90       	pop	r0
 31a:	0f 90       	pop	r0
 31c:	0f 90       	pop	r0
 31e:	0f 90       	pop	r0
 320:	df 91       	pop	r29
 322:	cf 91       	pop	r28

00000324 <.Loc.11>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 324:	eb cf       	rjmp	.-42     	; 0x2fc <sinf>
````

**Comment by salkinium** [comment]
I've also tested avr-gcc v11.3 and it doesn't have these issues. I'd therefore recommend waiting a bit more until GCC12 improves.

**Comment by salkinium** [comment]
Hm, comparing the listing from avr-gcc 11, it does call `sin` at the very bottom, while avr-gcc 12 calls `sinf` again.
```
0000032c <sinf>:
sinf():
./modm/ext/gcc/libstdc++/src/math.cc:24
  {
    return ::cos(x);
  }

  float sinf(float x)
  {
 32c:	cf 93       	push	r28
 32e:	df 93       	push	r29
 330:	00 d0       	rcall	.+0      	; 0x332 <L0^A>

00000332 <L0^A>:
 332:	0f 92       	push	r0
 334:	cd b7       	in	r28, 0x3d	; 61
 336:	de b7       	in	r29, 0x3e	; 62

00000338 <.Loc.8>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 338:	69 83       	std	Y+1, r22	; 0x01
 33a:	7a 83       	std	Y+2, r23	; 0x02
 33c:	8b 83       	std	Y+3, r24	; 0x03
 33e:	9c 83       	std	Y+4, r25	; 0x04
 340:	69 81       	ldd	r22, Y+1	; 0x01
 342:	7a 81       	ldd	r23, Y+2	; 0x02
 344:	8b 81       	ldd	r24, Y+3	; 0x03
 346:	9c 81       	ldd	r25, Y+4	; 0x04

00000348 <.LVL4>:
./modm/ext/gcc/libstdc++/src/math.cc:26
  }
 348:	0f 90       	pop	r0
 34a:	0f 90       	pop	r0
 34c:	0f 90       	pop	r0
 34e:	0f 90       	pop	r0
 350:	df 91       	pop	r29
 352:	cf 91       	pop	r28

00000354 <.Loc.11>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 354:	42 c4       	rjmp	.+2180   	; 0xbda <sin>
```

**Comment by salkinium** [comment]
Here is avr-gcc 10

```
0000032c <sinf>:
sinf():
./modm/ext/gcc/libstdc++/src/math.cc:24
  {
    return ::cos(x);
  }

  float sinf(float x)
  {
 32c:	cf 93       	push	r28
 32e:	df 93       	push	r29
 330:	00 d0       	rcall	.+0      	; 0x332 <L0^A>

00000332 <L0^A>:
 332:	0f 92       	push	r0
 334:	cd b7       	in	r28, 0x3d	; 61
 336:	de b7       	in	r29, 0x3e	; 62

00000338 <.Loc.8>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 338:	69 83       	std	Y+1, r22	; 0x01
 33a:	7a 83       	std	Y+2, r23	; 0x02
 33c:	8b 83       	std	Y+3, r24	; 0x03
 33e:	9c 83       	std	Y+4, r25	; 0x04
 340:	69 81       	ldd	r22, Y+1	; 0x01
 342:	7a 81       	ldd	r23, Y+2	; 0x02
 344:	8b 81       	ldd	r24, Y+3	; 0x03
 346:	9c 81       	ldd	r25, Y+4	; 0x04

00000348 <.LVL4>:
./modm/ext/gcc/libstdc++/src/math.cc:26
  }
 348:	0f 90       	pop	r0
 34a:	0f 90       	pop	r0
 34c:	0f 90       	pop	r0
 34e:	0f 90       	pop	r0
 350:	df 91       	pop	r29
 352:	cf 91       	pop	r28

00000354 <.Loc.11>:
./modm/ext/gcc/libstdc++/src/math.cc:25
    return ::sin(x);
 354:	42 c4       	rjmp	.+2180   	; 0xbda <sin>
```

**Comment by salkinium** [comment]
> It emits tons of warnings for register accesses, though:

https://wellys.com/posts/avr_c_gcc_error/

Apparently we need to add `–param=min-pagesize=0` to the ccflags. This fixes the warnings for my test builds.

**Comment by chris-durand** [comment]
> I'd therefore recommend waiting a bit more until GCC12 improves.

I don't think it's the compiler. Our older avr-gcc builds have avr-libc < 2.1.0 which defines the float variant math functions as macro aliases. In our libstdc++ we undefine the macros and replace them with real functions calling the double variants. In avr-libc 2.1.0 those are defined as functions now. So we have duplicate definitions and that is undefined behaviour. I would expect the linker to complain about this but it doesn't for `sinf`. We have merged a fix for that issue already but it did not cover all math functions.

@salkinium Could you try guarding all definitions in avr-libstdc++ math.cc with `#if defined(__AVR_LIBC_VERSION__) && (__AVR_LIBC_VERSION__ < 20100UL)` ?

**Comment by salkinium** [comment]
> Could you try guarding all definitions in avr-libstdc++ math.cc

Yup, that completely fixes it! Phew, fantastic, I was very confused as to how such a specific issue could be compiler related. PEBCAK as per usual. Ok, great, lemme do some more testing, since that problem derailed the AVR unittests completely.

**Comment by chris-durand** [comment]
Also, I am not sure it was a good idea to increase the OS version this builds on. On my work laptop I am currently stuck on Ubuntu 20.04 and I can't run the compiler there since it requires glibc 2.34 for no apparent technical reason. That issue will be there for everyone on any enterprise LTS distro like RHEL, CentOS and SuSE.

**Comment by salkinium** [comment]
After splitting up the AVR mega pro unittests a bit more (probably ran out of RAM again) and removing all `*f` math functions, all unittests now work again. So the compiler works fine.

**Comment by salkinium** [comment]
> That issue will be there for everyone on any enterprise LTS distro like RHEL, CentOS and SuSE.

I've reverted back to Ubuntu 20.04, that should fix it? We can always upgrade it later.

**Pull request merged**